pkg/api/mock/matcher/matcher_object.go

```diff
@@ -0,0 +1,76 @@
+package matcher
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+
+	gomegaTypes "github.com/onsi/gomega/types"
+	"go.anx.io/go-anxcloud/pkg/api/types"
+	"go.anx.io/go-anxcloud/pkg/utils/object/compare"
+)
+
+// Object matches APIObject with provided Object
+// It compares Objects by matchFields
+func Object(obj types.Object, matchFields ...string) gomegaTypes.GomegaMatcher {
+	return &objectMatcher{
+		obj:         obj,
+		matchFields: matchFields,
+	}
+}
+
+type objectMatcher struct {
+	obj         types.Object
+	matchFields []string
+}
+
+var (
+	// ErrObjectMatcherTypeMismatch is returned when ObjectMatcher is called with different Object types
+	ErrObjectMatcherTypeMismatch = errors.New("Provided Objects have different types")
+)
+
+func (m *objectMatcher) Match(actual interface{}) (bool, error) {
+	diff, err := m.compare(actual)
+	if err == ErrObjectMatcherTypeMismatch {
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+
+	return len(diff) == 0, nil
+}
+
+func (m *objectMatcher) FailureMessage(actual interface{}) string {
+	diff, err := m.compare(actual)
+	if err != nil {
+		return err.Error()
+	}
+	return fmt.Sprintf("ACTUAL Object does not match EXPECTED Object: %#v", diff)
+}
+
+func (m *objectMatcher) NegatedFailureMessage(actual interface{}) string {
+	diff, err := m.compare(actual)
+	if err != nil {
+		return err.Error()
+	}
+	return fmt.Sprintf("ACTUAL Object does match EXPECTED Object: %#v", diff)
+}
+
+func (m *objectMatcher) compare(actual interface{}) ([]compare.Difference, error) {
+	obj, err := castToAPIObjectPointer(actual)
+	if err != nil {
+		return nil, err
+	}
+
+	if reflect.TypeOf(m.obj) != reflect.TypeOf(obj.Unwrap()) {
+		return nil, ErrObjectMatcherTypeMismatch
+	}
+
+	diff, err := compare.Compare(obj.Unwrap(), m.obj, m.matchFields...)
+	if err != nil {
+		return nil, err
+	}
+
+	return diff, nil
```

Comment:
We currently get an output like this:
`ACTUAL Object does not match EXPECTED Object: []compare.Difference{compare.Difference{Key:"Name", A:"test", B:"foo"}}`

But with a simple change, we can get a much nicer output like this:
```
ACTUAL Object does not match EXPECTED Object:
  Key "Name"
  Expected
      <string>: test
  to equal
      <string>: foo
```

```golang
	ret := make([]string, 0, len(diff))
	for _, d := range diff {
		ret = append(ret, fmt.Sprintf(
			"Key %q\n%s\n\n",
			d.Key,
			format.Message(d.A, "to equal", d.B),
		))
	}

	return ret, nil
```

(combined with returning a `[]string` and using `strings.Join(diff, "\n")` in `FailureMessage()`.

Reply:
oh yeah, better do that in `FailureMessage`, we don't need to do that for `Match` or `NegatedFailureMessage` after all - good catch!